Support for early access to the operation and session task from a request

boxtcarpel opened the pull request:
Introducing a 'prepare' method for requests and operations, such that the
underlying session task could be accessed before an operation is enqueued
and contain the complete information.

boxcla commented:
Verified that @boxtcarpel has signed the CLA. Thanks for the pull request!

thuyBox commented on BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:
```diff
@@ -205,6 +205,23 @@ - (void)prepareAPIRequest
+                self.accessToken = self.session.accessToken;
+            }
+            NSError *error = nil;
+            self.sessionTask = [self createSessionTaskWithError:&error];
```
if there's an error creating session task, we should execute logic of line 374 to 377 to set self.error accordingly. We potentially have a valid self.sessionTask but there's other error setting up session task (in background tasks, we need to set up its cache before the task starts)

boxtcarpel replied:
Done.

Commit:
Support for early access to the operation and session task from a request

Introducing a 'prepare' method for requests and operations, such that the
underlying session task could be accessed before an operation is enqueued
and contain the complete information.

thuyBox commented on BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:
```diff
@@ -335,18 +358,18 @@ - (void)executeSessionTask
-            [self prepareAPIRequest];
-            self.accessToken = self.session.accessToken;
+    if (![self isCancelled]) {
+        if (self.sessionTask == nil) {
```
minor: not sure why we need this check explicitly?

boxtcarpel replied:
To avoid doing that work twice, now that it might actually get done on 'prepareOp'

thuyBox commented on BoxContentSDK/BoxContentSDK/Requests/BOXRequest.m:
```diff
@@ -78,6 +78,13 @@ - (BOXAPIOperation *)createOperation
+- (void)prepareOperation
+{
+    [self.operation.APIRequest setValue:[self userAgent] forHTTPHeaderField:@"User-Agent"];
+    [self.operation prepareOperation];
```
will we have a separate commit to expose sessionTask from BOXRequest?

boxtcarpel replied:
I don't think we need to... prepareOperation is exposed on request, and sessionTask is exposed on the operation that is exposed by the request.
Expected usage is like this:
           let request = BOXContentClient.default().fileUploadNewVersionRequestInBackground(withFileID:...)
            request?.requestAllFileFields = true
            request?.prepareOperation()
            if let task = request?.operation.sessionTask {
                NSFileProviderManager.default.register(task, ...

thuyBox commented:
+2

thuyBox commented:
+2
